### DISCUSSION TITLE
Add Streaming Option?

### DISCUSSION THREAD

**Discussion by vinookie** [PlexRipper/PlexRipper#399] [General]
First, thank you for your work on PlexRipper—it’s a fantastic tool! I was wondering if it would be possible to add a streaming capability to the program. Specifically, this feature would allow users to stream media directly while it’s being downloaded, rather than waiting for the download to complete.

The idea is to enable functionality where files are downloaded in a buffered manner and served via an HTTP endpoint for playback in media players or browsers. For example, implementing HTTP Range Requests or HLS could make this possible.

Is this something you would consider adding to PlexRipper? If not, could you share any guidance on how someone might go about extending the project to support such functionality?

**Comment by JasonLandbridge** [comment]
Hey, thank you very much! 

> I was wondering if it would be possible to add a streaming capability to the program. Specifically, this feature would allow users to stream media directly while it’s being downloaded, rather than waiting for the download to complete.

I honestly don't see the added value of a feature like this, and it might not be feasible. 

Since you have Plex Server ⇒ PlexRipper download, but then at the same time, PlexRipper needs to make the buffer available to stream. This buffer is currently written directly to disk, so there are two options here. 1, the buffer/stream is kept in memory, but this requires a large amount of memory on devices, or 2, it is read from disk again while at the same time it is writing to disk because it is downloading. This is incredibly slow and inefficient. And then we're not talking yet about what happens when the download has completed, should the stream then be cut and reloaded again from the final destination?

You could argue for a Web player that plays completed downloads, which should not be that challenging. However, Plex web player already does this, so I don't see this worth the effort 

> Is this something you would consider adding to PlexRipper? If not, could you share any guidance on how someone might go about extending the project to support such functionality?

My answer would be no on this one, mostly for the high amount of investment with little return. Technically you could re-open the media files after they have been downloaded, turn that into a stream API and then hook up a web-player to it. But again, you're just remaking Plex web player at that point. 

And PlexRipper will place the media files in the library that is watched by Plex, so it's like a 15-second wait before Plex has picked it up, and you can play it in Plex anyway.

**Comment by vinookie** [comment]
You’ve already touched on some valid challenges and trade-offs, but let me address your points from another perspective.

One possible advantage of implementing a streaming feature during download is addressing scenarios with limited server storage. Users could temporarily stream content directly while it downloads, freeing up space by deleting the file post-viewing if they don't wish to keep it. This system minimizes permanent storage usage while maintaining functionality for users with constrained capacity.

Using chunking to avoid excessive I/O overhead.

After the download is complete, the stream could seamlessly transition to the finalized file without interruption. This could involve a smart media handoff mechanism, which admittedly adds complexity but isn’t impossible.

While your point about Plex Web Player already serving this need is valid, this feature could cater to users who prefer managing their libraries differently (e.g., temporary, minimalistic setups). The cost-to-benefit ratio remains questionable, but it could appeal to a niche audience or unique use cases.

**Reply by JasonLandbridge** [reply]
Did you just use Chat-GPT to write this response for you?

![image](https://github.com/user-attachments/assets/b2eeb31d-5d8d-4f74-a9d3-df823e7b7438)

**Reply by vinookie** [reply]
Of course, I wanted to formulate my problem as precisely as possible for you.
